src/libs/datatableproviders/DataProvider.ts

```diff
@@ -0,0 +1,78 @@
+// define metadata structures
+interface EntityTypeMetadata {
+    attributeNames: string[],
+    count: number,
+    idName: string
+}
+
+export interface EntityMetadata {
+    [index: string]: EntityTypeMetadata
```

Comment:
I'm a ts newbie… can you explain what this syntax means?

Reply:
`[index: string]` means that this object has arbitrarily-named keys, but those keys must be strings. Look for "Index Signatures" here: https://www.w3schools.com/typescript/typescript_object_types.php